tools/gbagfx/jasc_pal.c

```diff
@@ -2,155 +2,87 @@
+// "<RED> <GREEN> <BLUE> <ALPHA (discarded if present)>" (color entry)
+//
+// Line endings can be \r\n or \n
 //
 // Each color component is a decimal number from 0 to 255.
 // Examples:
-// Black - "0 0 0\r\n"
-// Blue  - "0 0 255\r\n"
-// Brown - "150 75 0\r\n"
+// Black - "0 0 0"
+// Blue  - "0 0 255"
+// Brown - "150 75 0"
+// White  - "255 255 255 255"
+//                       ^~~ ignored
 
-#define MAX_LINE_LENGTH 11
+// 15 chars of color info, \r\n or \n, and \0
+#define MAX_LINE_LENGTH 18
 
-void ReadJascPaletteLine(FILE *fp, char *line)
+int ReadJascPaletteLine(FILE *fp, char *line)
 {
-    int c;
-    int length = 0;
-
-    for (;;)
-    {
-        c = fgetc(fp);
-
-        if (c == '\r')
-        {
-            c = fgetc(fp);
-
-            if (c != '\n')
-                FATAL_ERROR("CR line endings aren't supported.\n");
-
-            line[length] = 0;
-
-            return;
-        }
-
-        if (c == '\n')
-            FATAL_ERROR("LF line endings aren't supported.\n");
-
-        if (c == EOF)
-            FATAL_ERROR("Unexpected EOF. No CRLF at end of file.\n");
-
-        if (c == 0)
-            FATAL_ERROR("NUL character in file.\n");
-
-        if (length == MAX_LINE_LENGTH)
-        {
-            line[length] = 0;
-            FATAL_ERROR("The line \"%s\" is too long.\n", line);
-        }
-
-        line[length++] = c;
-    }
+    // Read line up to first newline char (inclusive) or [MAX_LINE_LENGTH-1]
+    if(fgets(line, MAX_LINE_LENGTH, fp) == NULL)
+        return 0;
+    line[strcspn(line, "\r\n")] = 0; // Terminate the line at the first newline char
+    return 1;
 }
 
 void ReadJascPalette(char *path, struct Palette *palette)
 {
-    char line[MAX_LINE_LENGTH + 1];
-
-    FILE *fp = fopen(path, "rb");
-
-    if (fp == NULL)
-        FATAL_ERROR("Failed to open JASC-PAL file \"%s\" for reading.\n", path);
+    char line_buffer[MAX_LINE_LENGTH];
+    int red, green, blue;
+    int numColors;
 
-    ReadJascPaletteLine(fp, line);
+    FILE *fp = fopen(path, "r");
 
-    if (strcmp(line, "JASC-PAL") != 0)
-        FATAL_ERROR("Invalid JASC-PAL signature.\n");
+    if (!fp)
+        FATAL_ERROR("Cannot open JASC-PAL file \"%s\" with error: %s\n", path, strerror(errno));
 
-    ReadJascPaletteLine(fp, line);
+    // Check JASC-PAL Header
+    ReadJascPaletteLine(fp, line_buffer);
+    if (strcmp(line_buffer, "JASC-PAL") != 0)
+        FATAL_ERROR("Invalid signature, expected \"JASC-PAL\", read: \"%s\"\n",line_buffer);
 
-    if (strcmp(line, "0100") != 0)
+    ReadJascPaletteLine(fp, line_buffer);
+    if (strcmp(line_buffer, "0100") != 0)
         FATAL_ERROR("Unsuported JASC-PAL version.\n");
 
-    ReadJascPaletteLine(fp, line);
+    // Get number of colors in palette
+    ReadJascPaletteLine(fp, line_buffer);
+    numColors = strtol(line_buffer, NULL, 10);
 
-    if (!ParseNumber(line, NULL, 10, &palette->numColors))
-        FATAL_ERROR("Failed to parse number of colors.\n");
```

Comment:
Revert `strtol` here to `ParseNumber` (it should error here if it fails to parse, rather than give an error later because the number of colors ended up as 0)

Reply:
Oh good point, I was also assigning a long int return value to a regular int in the `Palette` struct, `ParseNumber()` handles that as well